### PR TITLE
Allow addons to register fieldsets

### PR DESCRIPTION
**Pull request by edalzell**
Continuation of #3717 based on Jason's feedback

### CODE REVIEW

**Comment by aerni** [comment]
@morhi If you've got time for this ;-)

**Comment by edalzell** [comment]
The `exists` and `find` could do w/ a refactor I think, but that's for a later day

File: src/Fields/Fieldset.php [hunk 1]
```diff
@@ -25,7 +25,7 @@ public function handle(): ?string
         return $this->handle;
     }
 
-    public function setContents(array $contents)
+    public function setContents(array $contents): self
```
Comment (jasonvarga) [inline src/Fields/Fieldset.php:28]: Please revert everything in this file.

File: src/Providers/AddonServiceProvider.php [hunk 2]
```diff
@@ -11,32 +11,35 @@
...
+    protected $subscribe = [];
+    protected $tags = [];
     protected $updateScripts = [];
     protected $viewNamespace;
-    protected $publishAfterInstall = true;
+    protected $widgets = [];
+
     protected $config = true;
+    protected $publishAfterInstall = true;
     protected $translations = true;
```
Comment (jasonvarga) [inline src/Providers/AddonServiceProvider.php:43]: Please revert the re-ordering of these properties. Just include any new or removed properties.

File: src/Providers/AddonServiceProvider.php [hunk 3]
```diff
@@ -47,26 +50,27 @@ public function boot()
...
+                ->bootSchedule()
+                ->bootScopes()
+                ->bootScripts()
+                ->bootStylesheets()
+                ->bootTags()
+                ->bootTranslations()
                 ->bootUpdateScripts()
                 ->bootViews()
-                ->bootPublishAfterInstall();
+                ->bootWidgets();
```
Comment (jasonvarga) [inline src/Providers/AddonServiceProvider.php:73]: Please revert the re-ordering of these lines. Just include any new or removed methods.

Unless of course there's a reason for reordering.

**Review by jasonvarga — CHANGES_REQUESTED** [review]
Please revert the `phpunit.xml.dist -> phpunit.xml` rename.

File: src/Providers/AddonServiceProvider.php (hunk 3 above)
Comment (edalzell) [reply src/Providers/AddonServiceProvider.php:73]: > Unless of course there's a reason for reordering.

My sanity.

**Comment by edalzell** [comment]
@jasonvarga did those

**Comment by jasonvarga** [comment]
Hey Erin, I'm going to close this (sorry 😅) as we've given it some more thought to how it all should work in statamic/ideas#644. Please chime in over there.

❤️

**Pull request closed without merging**